test/extensions/filters/http/jwt_authn/group_verifier_test.cc

```diff
@@ -0,0 +1,537 @@
+
+  auto mock_auth = std::make_unique<MockAuthenticator>();
+  EXPECT_CALL(*mock_auth.get(), doVerify(_, _, _))
+      .WillOnce(Invoke([](Http::HeaderMap&, std::vector<JwtLocationConstPtr>*,
+                          AuthenticatorCallback callback) { callback(Status::Ok); }));
+  mock_auths_.push_back(std::move(mock_auth));
+
+  EXPECT_CALL(mock_cb_, onComplete(_)).WillOnce(Invoke([](const Status& status) {
+    ASSERT_EQ(status, Status::Ok);
+  }));
+  auto headers = Http::TestHeaderMapImpl{
+      {"sec-istio-auth-userinfo", ""},
+  };
+  context_ = Verifier::createContext(headers, &mock_cb_);
+  verifier_->verify(context_);
+  EXPECT_FALSE(headers.has("sec-istio-auth-userinfo"));
+}
+
+// require alls that just ends
+TEST_F(GroupVerifierTest, CanHandleUnexpectedEnd) {
+  const char config[] = R"(
+providers:
+  example_provider:
+    issuer: https://example.com
+    audiences:
+    - example_service
+    - http://example_service1
+    - https://example_service2/
+    remote_jwks:
+      http_uri:
+        uri: https://pubkey_server/pubkey_path
+        cluster: pubkey_cluster
+    forward_payload_header: sec-istio-auth-userinfo
+rules:
+- match: { path: "/match" }
+  requires:
+    requires_all:
+      requirements:
+      - requires_all:
+)";
+  MessageUtil::loadFromYaml(config, proto_config_);
+  createVerifier();
+
+  auto mock_auth = std::make_unique<MockAuthenticator>();
+  EXPECT_CALL(*mock_auth.get(), doVerify(_, _, _)).Times(0);
+  mock_auths_.push_back(std::move(mock_auth));
+
+  EXPECT_CALL(mock_cb_, onComplete(_)).WillOnce(Invoke([](const Status& status) {
+    ASSERT_EQ(status, Status::Ok);
+  }));
+  auto headers = Http::TestHeaderMapImpl{};
+  context_ = Verifier::createContext(headers, &mock_cb_);
+  verifier_->verify(context_);
+}
+
+TEST_F(GroupVerifierTest, TestRequiresAll) {
+  MessageUtil::loadFromYaml(RequiresAllConfig, proto_config_);
+
+  for (int i = 0; i < 2; ++i) {
+    auto mock_auth = std::make_unique<MockAuthenticator>();
+    EXPECT_CALL(*mock_auth.get(), doVerify(_, _, _))
+        .WillRepeatedly(Invoke([](Http::HeaderMap&, std::vector<JwtLocationConstPtr>*,
+                                  AuthenticatorCallback callback) { callback(Status::Ok); }));
+    mock_auths_.push_back(std::move(mock_auth));
+  }
+  createVerifier();
+
+  EXPECT_CALL(mock_cb_, onComplete(_)).WillOnce(Invoke([](const Status& status) {
+    ASSERT_EQ(status, Status::Ok);
+  }));
+  auto headers = Http::TestHeaderMapImpl{
+      {"example-auth-userinfo", ""},
+      {"other-auth-userinfo", ""},
+  };
+  context_ = Verifier::createContext(headers, &mock_cb_);
+  verifier_->verify(context_);
+  EXPECT_FALSE(headers.has("example-auth-userinfo"));
+  EXPECT_FALSE(headers.has("other-auth-userinfo"));
+}
+
+// test requires all with first auth returning bad format
+TEST_F(GroupVerifierTest, TestRequiresAllBadFormat) {
+  MessageUtil::loadFromYaml(RequiresAllConfig, proto_config_);
+  std::vector<AuthenticatorCallback> callbacks;
```

Comment:
I think "callbacks" vector and creating a list of mock_auths_ can be moved into a function of class GroupVerifierTest as
createMockAuths(int n);   so it can be shared by multiple tests.